tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+defaultCountFields = countFields1 . from
+
+instance CountFields1 V1 where
+  countFields1 _ = 0
+
+instance CountFields1 U1 where
+  countFields1 _ = 0
+
+instance CountFields1 (K1 i c) where
+  countFields1 _ = 1
+
+instance CountFields1 f => CountFields1 (M1 i c f) where
+  countFields1 (M1 x) = countFields1 x
+
+instance (CountFields1 a, CountFields1 b) => CountFields1 (a :+: b) where
+  countFields1 (L1 x) = countFields1 x
+  countFields1 (R1 x) = countFields1 x
+
+instance (CountFields1 a, CountFields1 b) => CountFields1 (a :*: b) where
+  countFields1 (a :*: b) = countFields1 a + countFields1 b
+```
+
+You could notice that some data types like `Par0`, `Rec1` didn't get their
+definitions. This is OK because we work with `Generic`, not `Generic1` here,
+but omitting definitions for some data types that generic representations
+are built from is normal. As `GHC.Generics` docs say:
+
+* If no `:+:` instance is given, the function may still work for empty
+  datatypes or datatypes that have a single constructor, but will fail on
+  datatypes with more than one constructor.
+
+* If no `:*:` instance is given, the function may still work for datatypes
+  where each constructor has just zero or one field, in particular for
+  enumeration types.
+
+* If no `K1` instance is given, the function may still work for enumeration
+  types, where no constructor has any fields.
+
+* If no `V1` instance is given, the function may still work for any datatype
+  that is not empty.
+
+* If no `U1` instance is given, the function may still work for any datatype
+  where each constructor has at least one field.
+
+An `M1` instance is always required (but it can just ignore the
+meta-information).
+
+## Putting the magic into the type classes
+
+Having dealt with generic implementation of the functionality of interest,
+let's put it all together and use a special GHC extension to allow the user
+derive type classes without knowing anything about generics.
+
+For generic implementation to work without user's definition we need to
+provide it as the default definition. As you have already seen, generic
+implementation often involves `Generic` constraints and it would be ugly to
+add them to every type class (on the left hand side of `=>`) just to make
+deriving easier. The `default` keyword, enabled by the `DefaultSignartures`
+language extension, allows to give a different type signature for default
+implementation of a type class:
+
+```haskell
+instance Functor f where
+  fmap :: (a -> b) -> f a -> f b
+  default fmap :: (Generic1 f, Functor (Rep1 f)) => (a -> b) -> f a -> f b
+  fmap = to1 . fmap . from1
+
+class CountFields a where
+  countFields :: a -> Natural
+  default countFields :: (Generic a, CountFields1 (Rep a)) => a -> Natural
+  countFields = defaultCountFields
+```
+
+This way we can eat the cake and have it: deriving is easy and otherwise no
+ugly details are visible!
+
+## Conclusion
+
```

Comment:
Consider rephrasing as:
Generics are a powerful means of automating the process of defining type class instances, which is error-prone and boring if done by hand.